Refactor of internal role hierarchy model + changes for 3.0.0

ezimuel opened the pull request:
This PR contains a big refactoring of the role hierarchy model with minor changes in the APIs. This PR contains a potential 3.0.0 release for `zend-permissions-rbac`, containing BC breaks.

This PR includes the following changes:
### Added

- Check for circular references in role hierarchy when using `Role::addChild()` and `Role::addParent()` functions.

### Changed

- `Role::addChild(RoleInterface $child)`, accepts only `RoleInterface` parameter, no string anymore.
- `Role::addParent(RoleInterface $parent)`, accepts only `RoleInterface` parameter, no string anymore.
- `Zend\Permissions\Rbac\AssertionInterface`, added the optional parameters `$permission` and `$role` in the `assert()` function, as follows:
  `assert(Rbac $rbac, string $permission = null, RoleInterface $role = null)`

### Deprecated

- Nothing.

### Removed

- [AbstractIterator](https://github.com/zendframework/zend-permissions-rbac/blob/release-2.6.0/src/AbstractIterator.php), removed the class. The role hierarchy is not based anymore on `RecursiveIterator`.

- [AbstractRole](https://github.com/zendframework/zend-permissions-rbac/blob/release-2.6.0/src/AbstractRole.php),  removed the class. All the functions have been moved in `Zend\Permissions\Rbac\Role`.

- `Role::setParent()`, use `Role::addParent()` instead.

### Fixed

- [#30](https://github.com/zendframework/zend-permissions-rbac/issues/30), Fixed circular references with the protected functions `Role::hasAncestor($role)` used in `Role::addChild()` and `Role::hasDescendant($role)` in `Role::addParent()`.

webimpress commented:
@ezimuel just a quick question - are we not going to bump PHP requirements to version 7.1 with 3.0.0 release??

ezimuel commented:
@webimpress yes, I'm removing the support for PHP 5.6 and PHP 7.0, thanks for the reminder.

ezimuel commented:
@webimpress fixed now, requiring PHP 7.1+.

ezimuel commented:
Ping @Ocramius, @webimpress for review.
Ping also @olegkrivtsov, @asgrim, @DominicDetta, @codeaid for comments and feedback.

webimpress commented on composer.json:
```diff
@@ -3,7 +3,7 @@
     "license": "BSD-3-Clause",
     "keywords": [
-        "zf2",
+        "zf3",
```
We stop using version with zf components, so I'd suggest to use only `zf` and add please `zendframework` (please see maintainers repository and templates there)

webimpress commented on composer.json:
```diff
@@ -29,7 +29,7 @@
     },
     "require-dev": {
-        "phpunit/phpunit": "^5.7.15|| ^6.2.1",
+        "phpunit/phpunit": "^6.4.4",
```
We can switch to PHPUnit 7.0.1 as it supports only PHP 7.1+ !

webimpress commented on doc/book/examples.md:
```diff
@@ -98,7 +98,7 @@ class AssertUserIdMatches implements AssertionInterface
     }
 
-    public function assert(Rbac $rbac)
+    public function assert(Rbac $rbac, $permission = null, $role = null)
```
Can we add some typehints? If these types are mixed could we please add PHPDocs about allowed types?

webimpress commented on src/Assertion/CallbackAssertion.php:
```diff
@@ -31,14 +32,9 @@ public function __construct($callback)
-     *
-     * @param Rbac $rbac
-     * @return bool
+     * {@inheritDoc}
```
This comment doesn't have any value, I would remove it.

webimpress commented on src/Assertion/CallbackAssertion.php:
```diff
@@ -31,14 +32,9 @@ public function __construct($callback)
+     * {@inheritDoc}
      */
-    public function assert(Rbac $rbac)
+    public function assert(Rbac $rbac, string $permission = null, RoleInterface $role = null)
```
Can we add also return type in function declaration, please?

webimpress commented on src/AssertionInterface.php:
```diff
@@ -14,8 +14,10 @@
      * @return bool
      */
-    public function assert(Rbac $rbac);
+    public function assert(Rbac $rbac, string $permission = null, RoleInterface $role = null);
```
Can we have here also return type declaration in function signature (`bool`)?

webimpress commented on src/AssertionInterface.php:
```diff
@@ -14,8 +14,10 @@
-     * @param  Rbac    $rbac
+     * @param  Rbac $rbac
+     * @param  string $permission
+     * @param  RoleInterface $role
```
This PHPDoc can be removed, as everything is specified in the function signature, so these information are redundant.

webimpress commented on src/Exception/ExceptionInterface.php:
```diff
@@ -3,7 +3,7 @@
  *
  * @link      http://github.com/zendframework/zf2 for the canonical source repository
- * @copyright Copyright (c) 2005-2015 Zend Technologies USA Inc. (http://www.zend.com)
+ * @copyright Copyright (c) 2005-2018 Zend Technologies USA Inc. (http://www.zend.com)
```
As I remember we update date only when we modify something in the file. Here nothing is changed so the year shouldn't be changed.

froschdesign replied:
Where is this requirement defined?

webimpress commented on src/Exception/RuntimeException.php:
```diff
@@ -0,0 +1,15 @@
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2018 Zend Technologies USA Inc. (http://www.zend.com)
```
For new files we should use only current year.

froschdesign replied:
Where is this requirement defined?

webimpress commented on src/Exception/RuntimeException.php:
```diff
@@ -0,0 +1,15 @@
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
```
I think it would be nice to update all headers to point to the repository (as we have in many others repos already, for example see any expressive repo)

webimpress commented on src/Exception/RuntimeException.php:
```diff
@@ -0,0 +1,15 @@
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2018 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   http://framework.zend.com/license/new-bsd New BSD License
```
The same here, this link doesn't even work....

webimpress commented on src/Rbac.php:
```diff
@@ -22,20 +25,19 @@ class Rbac extends AbstractIterator
      * @return \Zend\Permissions\Rbac\Rbac
      */
-    public function setCreateMissingRoles($createMissingRoles)
+    public function setCreateMissingRoles(bool $createMissingRoles): Rbac
```
We are using single space before colon in return type declaration.

webimpress replied:
Also as a return type I would use `self` (it's clearer when you read class in the middle then we know straight away that we return the same class).

weierophinney replied:
Or we should drop the fluent interface altogether, as we have in other components, and return `void`. (This comment applies to all setter methods that currently implement a fluent interface.)

ezimuel replied:
I removed fluent interface in all the library.

webimpress commented on src/Rbac.php:
```diff
@@ -22,20 +25,19 @@ class Rbac extends AbstractIterator
     /**
-     * @param  bool                     $createMissingRoles
+     * @param  bool $createMissingRoles
      * @return \Zend\Permissions\Rbac\Rbac
```
Imho this should be updated to `return $this`

codeaid commented:
Hi all,

Thanks for including me in this conversation!

As you can imagine, a lot has happened since I posted the original issue. I ended up developing my own solution, which, to be honest, I'm really happy with and having successfully used it for several months in production environment there is no real need for me to switch back to Zend RBAC.

My solution literally consists of two classes - container and role - and supports adding roles by names and works around circular references. I posted the related files in [this Gist](https://gist.github.com/codeaid/013bbd8a0e398d0f9e2db5ed3c2f486d) in case anyone's interested to have a look. There's a `usage.php` file at the bottom showing how I'm using the container by building it from my service, which could really be anything.

Speaking about the PR - looks good to me at a glance. I like that the `add*` methods use type hinting -
 stricter is always better in my books. Also, getting rid of the iterators should result in some performance increase (even if it's a small one). It's always good to strip out something that is not absolutely needed.

Other than that, I can only go by what @weierophinney says as he's the expert when it comes to framework standards. Sorry, if I'm not really being helpful :/

froschdesign commented:
@codeaid 
> …there is no real need for me to switch back to Zend RBAC.

Maybe (unit) tests? 😉

codeaid commented:
@froschdesign You mean unit testing my implementation or `zend-permissions-rbac`?

ezimuel commented:
@codeaid thanks for sharing your code with us. It's always interesting to compare different solutions to the same problem. If this works for you that's great! Of course, as @froschdesign suggested unit test will help. I suggest to add it in your implementation to find issues before they will happens in production. Thanks again for your feedbacks!

codeaid commented:
@froschdesign @ezimuel - Of course I have unit tests! How could you even assume otherwise!? :D I added them to the gist (if it helps). I believe they yield a 100% coverage too so I'm covered (pun intended).

ezimuel commented:
@webimpress, @weierophinney I incorporated all the feedbacks, including: `declare(strict_types=1);` on each file, and removing fluent interface. Moreover, I added some unit tests to improve the code coverage.

ezimuel commented:
Now we have 100% code coverage!

DominicDetta commented:
Ciao a tutti 😄 
I'm looking forward to this release because it contains changes which are perfect for my requirements.
I really like the new signature of the `assert()` function which by the way is something similar to what I did currently in a project.  I'm not an expert so I don't think I can be useful, but I can express some questions.
Right now there have been only 2 requirements that I don't know how to achieve with this component:

- How to deny  the inheritance for a specific permission (more explanation [here](https://discourse.zendframework.com/t/zend-permissions-rbac-inheritance-permissions/462))
- How to return to the client a list of resources filtered by the RBAC rules (I would need to invoke the `isGranted ` method for each resource to verifiy accessibility by the current role: the permission string need to be generated for each resource and eventually the assert class can be different for each resource)

As always thanks for your contributions 👍

weierophinney commented on composer.json:
```diff
@@ -13,14 +15,14 @@
-            "dev-master": "2.6-dev",
-            "dev-develop": "2.7-dev"
+            "dev-master": "3.0-dev",
+            "dev-develop": "3.1-dev"
```
Don't change these until the develop branch is merged to master (i.e., right before the 3.0 release).

When you do, use the format `3.0.x-dev` and `3.1.x-dev` (this is the recommended format by Composer).

weierophinney commented on CHANGELOG.md:
```diff
@@ -2,49 +2,42 @@
+   parameter, no string anymore.
+- `Zend\Permissions\Rbac\AssertionInterface`, added the optional parameters
+  `$permission` and `$role` in the `assert()` function, as follows:
+  `assert(Rbac $rbac, string $permission = null, RoleInterface $role = null)`
```
Is the permission more likely than the role to be passed? I would have assumed this would be in the following order:

```php
assert(Rbac $rbac, RoleInterface $role = null, string $permission = null)
```

For that matter: are the permission and role arguments actually optional? Will there ever be a case where they are not provided?

ezimuel replied:
@weierophinney I added `$role` and `$permission` as optional parameters because in this way we can limit BC breaks from the previous implementation. The actual interface is:
```php
namespace Zend\Permissions\Rbac;
interface AssertionInterface
{
    /**
     * Assertion method - must return a boolean.
     *
     * @param  Rbac    $rbac
     * @return bool
     */
    public function assert(Rbac $rbac);
}
```

ezimuel replied:
@weierophinney I changed the order of params, using the follows:

```php
assert(Rbac $rbac, RoleInterface $role = null, string $permission = null)
```

weierophinney commented on doc/book/methods.md:
```diff
@@ -1,62 +1,39 @@
-`assert(Rbac $rbac) : bool` | Given an RBAC, determine if permission is granted.
+Method signature                                                            | Description
+--------------------------------------------------------------------------- | -----------
+`assert(Rbac $rbac, string $permission = null, RoleInterface $role = null)` | Given an RBAC, an optional permission, and optional role determine if permission is granted.
```
This somewhat answers my earlier question. My reading of this, though is that:

- The rbac is always passed
- The permission will likely always be passed (as we're asserting a permission)
- Shouldn't the role always be passed? Aren't we asserting the role has the given permission?

ezimuel replied:
Read my previous comment.

weierophinney commented on src/Assertion/CallbackAssertion.php:
```diff
@@ -31,14 +33,9 @@ public function __construct($callback)
-    public function assert(Rbac $rbac)
+    public function assert(Rbac $rbac, string $permission = null, RoleInterface $role = null): bool
     {
         return (bool) call_user_func($this->callback, $rbac);
```
Two things:

-  This should use the `($this->callback)()` form, as any callback can be called that way in the PHP 7 series.
- Should _all_ arguments be  passed to the callback?

ezimuel replied:
I changed it using `$this->callback()` and passing all the arguments.

weierophinney commented on src/Rbac.php:
```diff
@@ -22,20 +25,18 @@ class Rbac extends AbstractIterator
+     * @return void
      */
-    public function setCreateMissingRoles($createMissingRoles)
+    public function setCreateMissingRoles(bool $createMissingRoles): void
```
For consistency with other packages, and until we finish up zend-coding-standard, please add a space on either side of the `:` when specifying the return type hint: `) : void`. This applies to all RTH added in this patch.

ezimuel replied:
Done

codeaid replied:
@weierophinney  I know it's not my place to speak but I don't think there should be a space on the left side of the colon. All php.net code snippets (e.g. [here](http://php.net/manual/en/migration71.new-features.php)) have it without it and many IDE's (such as PhpStorm) default to not having it too.

Having a space to the left would mean extra configuration would need to be specified only to achieve something that's contrary to the current "standard".

weierophinney commented on src/Rbac.php:
```diff
@@ -67,65 +68,52 @@ public function addRole($child, $parents = null)
+        }
+        $roleName = $role->getName();
+        return isset($this->roles[$roleName]) &&
+               $this->roles[$roleName] === $role;
```
Push the `&&` to the next line, and indent only one additional level (4 spaces).

ezimuel replied:
Done

weierophinney commented on src/Rbac.php:
```diff
@@ -134,27 +122,32 @@ public function getRole($objectOrName)
+        }
+        if (is_callable($assert)) {
+            return $result && $assert($this, $permission, $role);
+        }
```
Considering there's already the `CallbackAssertion`, it might be easier to declare:

```php
AssertionInterface $assert = null
```

and require developers to decorate callable assertions before passing them to this method.

ezimuel replied:
I think we can limit the BC breaks here if we assume that `$assert` can be `AssertionInterface` or `Callable` or `null`, like the actual implementation.

weierophinney commented on src/Rbac.php:
```diff
@@ -134,27 +122,32 @@ public function getRole($objectOrName)
+            return $result;
+        }
+        if ($assert instanceof AssertionInterface) {
+            return $result && $assert->assert($this, $permission, $role);
```
This validates my questions from above regarding the `AssertionInterface`: the permission and role _are_ always passed here. As such, I think  you should make them required arguments, and change the order of the two.

ezimuel replied:
Again, read my previous comments on `AssertionInterface` optional params.

weierophinney commented on src/Role.php:
```diff
@@ -1,21 +1,183 @@
+            throw new Exception\RuntimeException(sprintf(
+                "To prevent circular references, you cannot add role '%s' as parent",
+                $parentName
+            ));
```
Please create a new exception type for this condition, such as `CircularReferenceException`.

ezimuel replied:
Done

ezimuel commented:
@DominicDetta for the first question I replied on [discourse.zendframework.com](https://discourse.zendframework.com/t/zend-permissions-rbac-inheritance-permissions/462/2).

For the second question, if a **resource** that you mentioned is a **permission**, you should have a list of permission stored somewhere in your implementation (maybe into a db table or a configuration file). You can iterate on it and check for grant using the `Zend\Permission\Rbac\Rbac::isGranted()` function.

ezimuel commented:
@weierophinney I updated the PR with all your suggestions, thanks!

weierophinney commented:
The point is consistency within our components. We have been using the
spaces around the `:` operator consistently since we first started
declaring return types. If we do differently here, we are indicating
"anything goes".

Interestingly, there is an open discussion on psr-12 (which is a revision
of psr-1/2 for PHP 7) currently happening about this very subject; a number
of people would prefer the extra space.

For now: we need to keep a consistent style across all repos, which means
the extra space. If psr-12 goes in favor of no space, we'll revisit later
and update.

On Feb 19, 2018 11:01 AM, "codeaid" <notifications@github.com> wrote:

> *@codeaid* commented on this pull request.
> ------------------------------
>
> In src/Rbac.php
> <https://github.com/zendframework/zend-permissions-rbac/pull/34#discussion_r169131690>
> :
>
> >       */
> -    public function setCreateMissingRoles($createMissingRoles)
> +    public function setCreateMissingRoles(bool $createMissingRoles): void
>
> @weierophinney <https://github.com/weierophinney> I know it's not my
> place to speak but I don't think there should be a space on the left side
> of the colon. All php.net code snippets (e.g. here
> <http://php.net/manual/en/migration71.new-features.php>) have it without
> it and many IDE's (such as PhpStorm) default to not having it too.
>
> Having a space to the left would mean extra configuration would need to be
> specified only to achieve something that's contrary to the current
> "standard".
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendframework/zend-permissions-rbac/pull/34#discussion_r169131690>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABlVxknssjSMNzg0O2IJe-4F5JTnM9Yks5tWaj3gaJpZM4SEFws>
> .
>

weierophinney commented on src/AssertionInterface.php:
```diff
@@ -1,21 +1,18 @@
-     * @return bool
      */
-    public function assert(Rbac $rbac);
+    public function assert(Rbac $rbac, RoleInterface $role = null, string $permission = null) : bool;
```
I'm still not entirely convinced these arguments should be optional.

While I understand the argument of keeping BC breaks to a minimum, I also think we could address this in a new v2.X minor release by doing the following:

- Updating the `Rbac` class to pass the role and permission when calling an assertion.
- Documenting in migration notes to users that they should add these arguments to their assertions in order to prepare for the v3 release. Their assertions can ignore them for now — or even update their logic to make use of them!

Taking those steps allows the v3 release to make these required, as they are always passed.

ezimuel replied:
@weierophinney ok, this plan sounds good to me.

weierophinney replied:
In talking with @ezimuel today, we realized that for anybody implementing the `AssertionInterface`, they will need to make a signature change when upgrading to version 3 regardless of any changes we'd make in the v2 release branch. As such, we will not do another v2 release, and move ahead instead with the v3 release.

ezimuel commented:
@weierophinney I wrote the migration guide, removed the optionals in AssertionInterface and fixed an issue on CHANGELOG.

weierophinney commented:
@ezimuel I've edited and expanded the migration document; I think this is good to go at this point!

froschdesign commented on doc/book/methods.md:
```diff
@@ -1,62 +1,39 @@
 
 Method signature                                                            | Description
 --------------------------------------------------------------------------- | -----------
 `addRole(string|RoleInterface $child, array|RoleInterface $parents = null)` | Add a role to the RBAC. If `$parents` is non-null, the `$child` is also added to any parents provided.
```
Backslash is required for pipe:

`addRole(string\|RoleInterface $child, array\|RoleInterface $parents = null)`

froschdesign replied:
Problem is the rendering inside tables.

froschdesign commented on doc/book/methods.md:
```diff
@@ -1,62 +1,39 @@
 `addRole(string|RoleInterface $child, array|RoleInterface $parents = null)` | Add a role to the RBAC. If `$parents` is non-null, the `$child` is also added to any parents provided.
-`getRole(string|RoleInterface $role) : RoleInterface`                       | Recursively queries the RBAC for the given role, returning it if found, and raising an exception otherwise.
+`getRole(string $role) : RoleInterface`                                     | Get the role specified by name, raising an exception if not found.
 `hasRole(string|RoleInterface $role) : bool`                                | Recursively queries the RBAC for the given role, returning `true` if found, `false` otherwise.
```
Same here:

`hasRole(string\|RoleInterface $role) : bool`

froschdesign commented on doc/book/methods.md:
```diff
@@ -1,62 +1,39 @@
 `hasRole(string|RoleInterface $role) : bool`                                | Recursively queries the RBAC for the given role, returning `true` if found, `false` otherwise.
 `getCreateMissingRoles() : bool`                                            | Retrieve the flag that determines whether or not `$parent` roles are added automatically if not present when calling `addRole()`.
 `setCreateMissingRoles(bool $flag) : void`                                  | Set the flag that determines whether or not `$parent` roles are added automatically if not present when calling `addRole()`.
 `isGranted(string|RoleInterface $role, string $permission, $assert = null)` | Determine if the role has the given permission. If `$assert` is provided and either an `AssertInterface` instance or callable, it will be queried before checking against the given role.
```
Same here:

`isGranted(string\|RoleInterface $role, string $permission, $assert = null)`

ezimuel commented:
@froschdesign I added the backslash for pipe in docs, as you suggested. Thanks!

weierophinney commented:
Thanks, @ezimuel - great improvement! Merged to develop.
